docs/content/preview/architecture/transactions/isolation-levels.md

```diff
@@ -67,11 +67,49 @@ The following matrix shows conflicts between locks of different types at a high
 
-Further distinction exists between locks acquired on a DocDB node that is being written to by any
-transaction or read by a read-modify-write serializable transaction, and locks acquired on its parent nodes. The former types of locks are referred to as strong locks and the latter as weak locks. For example, if a snapshot isolation transaction is setting column `col1` to a new value in row `row1`, it acquires a weak snapshot isolation write lock on `row1` but a strong snapshot isolation write lock on `row1.col1`. Because of this distinction, the full conflict matrix actually looks more complex, as follows:
+Although described here as a separate lock type for simplicity, the snapshot isolation write lock type is actually implemented internally as a combination of the other two lock types.  That is, taking a single snapshot isolation write lock is equivalent to taking both a serializable read lock and a serializable write lock.
+
+## Locking granularities
+
+Locks can be taken at many levels of granularity.  For example, a serializable read lock could be taken at the level of an entire tablet, a single row, or a single column of a single row.  Such a lock will block attempts to take write locks at that or lower granularities. Thus, for example, a read lock taken at the row level will block attempts to write to that entire row or any sub column of it.
+
+In addition to the above-mentioned levels of granularity, locks in DocDB can be taken at prefixes of the primary key, treating the hash columns as a single unit.  For example, if we have created a YSQL table via:
+```sql
+CREATE TABLE test (h1 INT, h2 INT, r1 INT, r2 INT, v INT w INT PRIMARY KEY ((h1,h2) HASH, r1 ASC, r2 ASC);
+```
+
+then we could lock at any of the following:
+- the entire tablet
+- all rows having h1=2, h2=3
+- all rows having h1=2, h2=3, r1=4
+- the row having h1=2, h2=3, r1=4, r2=5
+- column v of the row having h1=2, h2=3, r1=4, r2=5
+
+With YCQL, granularities exist below the column level; for example only one key of a column map can be locked.
+
+## Efficiently detecting conflicts between locks of different granularities
+
+The straightforward way to handle locks of different granularities would be to have a map from location to lock types.  However, this is too inefficient for detecting conflicts: attempting, for example, to add a lock at the tablet level would require checking for locks at every row and column in that tablet.
+
+To make conflict detection efficient, YugabyteDB stores extra information at each location about the locks at lower granularities below it.  In particular, it takes a normal lock at the original granularity and weaker versions of that lock at all the granularities that enclose the original granularity.  We call the normal locks _strong_ locks and the weaker variants _weak_ locks.
```

Comment:
```suggestion
To make conflict detection efficient, YugabyteDB stores extra information at each location about the locks at granularities below it.  In particular, it takes a normal lock at the original granularity and weaker versions of that lock at all the granularities that enclose the original granularity. The normal locks are called _strong_ locks and the weaker variants _weak_ locks.
```

Reply:
style guide says to prefer active voice

Reply:
good: removing "lower", however

Reply:
Style guide not mention we don't use we, I forget. But we don't.

Reply:
:+1: